Shows photos in recycler view with pagination & full screen photo with download and share feature

jhnaiem opened the pull request.

Commit:
Create README.md

(cherry picked from commit e1fc8f3e74a901e65540f5e05e3cb3083bb846ab)

jhasan95 committed:
feat: added data layer to get photos from the API

- added retrofit dependency

jhasan95 committed:
refactor: added get photo by id in the data layer to get a single photo

jhasan95 committed:
feat: added domain layer with repo and usecase

jhasan95 committed:
feat: added layouts for gallery fragment, recyclerview item

- added navigation
- added data binding dependency
- added initial ui layer

jhasan95 committed:
feat: added viewmodel and adapter to show photos in the recyclerview

- Added internet and network permission in the manifest

jhasan95 committed:
feat: added common classes to load image using glide

- added hilt dependency
- adde glide dependency

jhasan95 committed:
feat: added di layer and annotated with hilt components

- added hilt dependency
- updated item photo layout

jhasan95 committed:
refactor: updated to item photo layout to show image in full screen

jhasan95 committed:
feat: added pagination & refactor: modified DI implementation

jhasan95 committed:
refactor: implemented getRefreshKey()

- removed unnecessary code

jhasan95 committed:
feat: added single photo viewer on tap on a photo

atiqaraj commented on app/build.gradle:
```diff
@@ -29,15 +32,55 @@ android {
-    implementation 'androidx.core:core-ktx:1.6.0'
-    implementation 'androidx.appcompat:appcompat:1.3.1'
+    implementation 'androidx.legacy:legacy-support-v4:1.0.0'
+    def nav_version = "2.3.5"
```
create a project level array and define variables there 
```
   ext.versions = [
            
    ]
```

atiqaraj commented on app/build.gradle:
```diff
@@ -29,15 +32,55 @@ android {
+    //Paging Library
+    implementation 'androidx.paging:paging-runtime-ktx:3.1.0'
+
 }
```
add new line on all files.

atiqmlbd commented on app/src/main/AndroidManifest.xml:
```diff
@@ -2,15 +2,19 @@
+
     <application
+        android:name=".UnsplashApp"
         android:allowBackup="true"
```
add support for Android 12

atiqaraj commented on app/src/main/java/com/example/unsplashphotos/common/ImageLoader.kt:
```diff
@@ -0,0 +1,67 @@
+import com.bumptech.glide.Glide
+
+
+class ImageLoader {
```
make it singleton using using hilt provided context

atiqaraj commented on app/src/main/java/com/example/unsplashphotos/data/Constants.kt:
```diff
@@ -0,0 +1,4 @@
+package com.example.unsplashphotos.data
+
+//const val BASE_URL = "https://api.unsplash.com/"
+const val CLIENT_ID = "RUvTzcL-6uXdrLTWpudHjLa9MqIvBcjPmJ8bqJwFpQg"
```
transfer this into build.gradle file and set up flavours

atiqaraj commented on app/src/main/java/com/example/unsplashphotos/data/RetrofitInstance.kt:
```diff
@@ -0,0 +1,32 @@
+import retrofit2.Retrofit
+import retrofit2.converter.gson.GsonConverterFactory
+
+class RetrofitInstance {
```
no need of this file. expose okhttp and retrofit instances using provides in Hilt

atiqaraj commented on app/src/main/java/com/example/unsplashphotos/data/api/PhotoService.kt:
```diff
@@ -0,0 +1,17 @@
+interface PhotoService {
+
+    @GET("/photos/")
+    suspend fun getPhotos(@Query("client_id") client_id: String,@Query("page") page: Int): Response<List<Photo>>
```
variable naming convention is not right

atiqaraj commented on app/src/main/java/com/example/unsplashphotos/data/api/PhotoService.kt:
```diff
@@ -0,0 +1,17 @@
+    suspend fun getPhotos(@Query("client_id") client_id: String,@Query("page") page: Int): Response<List<Photo>>
+
+    @GET("/photos/{id}")
+    suspend fun getPhotoById(@Path(value = "id") id: String, @Query("client_id") client_id: String): Response<Photo>
```
the same

atiqaraj commented on app/src/main/java/com/example/unsplashphotos/data/repository/PhotoExtension.kt:
```diff
@@ -0,0 +1,16 @@
+import com.example.unsplashphotos.data.model.Photo
+
+
+val DIFF_ITEM_CALLBACK = object : DiffUtil.ItemCallback<Photo>() {
```
make it a static variable of the adapter

atiqaraj commented on app/src/main/java/com/example/unsplashphotos/data/repository/PhotoPagingSource.kt:
```diff
@@ -0,0 +1,36 @@
+import javax.inject.Singleton
+
+@Singleton
+class PhotoPagingSource @Inject constructor(private val photoRepo: PhotoRepoImpl) : PagingSource<Int, Photo>() {
```
don't expose the implementation

atiqaraj commented on app/src/main/java/com/example/unsplashphotos/data/repository/PhotoRepoImpl.kt:
```diff
@@ -0,0 +1,57 @@
+        try {
+            val response = photoDataSource.getPhotoById(photoId)
+            val body = response.body()
+            Log.i("T===>", "Photo: " + body)
```
install timber

atiqaraj commented on app/src/main/java/com/example/unsplashphotos/data/repository/PhotoRepoImpl.kt:
```diff
@@ -0,0 +1,57 @@
+            Log.e("ErrorPhoto", exception.message.toString())
+        }
+
+        return photoStore!!
```
handle null state

atiqaraj commented on app/src/main/java/com/example/unsplashphotos/di/UseCaseModule.kt:
```diff
@@ -0,0 +1,22 @@
+
+@InstallIn(SingletonComponent::class)
+@Module
+class UseCaseModule {
```
no need to introduce UseCaseModule

atiqaraj commented on app/src/main/java/com/example/unsplashphotos/domain/usecase/PhotoUseCase.kt:
```diff
@@ -0,0 +1,22 @@
+
+    fun fetchPhotos(): Flow<PagingData<Photo>> {
+        return Pager(
+            PagingConfig(20, enablePlaceholders = false)
```
make it a constant

atiqaraj commented on app/src/main/java/com/example/unsplashphotos/ui/GalleryFragment.kt:
```diff
@@ -0,0 +1,76 @@
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        photoAdapter = PhotoAdapter{ selectedPhoto: Photo, extra ->
```
move this code to `setupPhotoRecyclerView`

atiqaraj commented on app/src/main/java/com/example/unsplashphotos/ui/GalleryFragment.kt:
```diff
@@ -0,0 +1,76 @@
+    private fun displayPhotos() {
+        lifecycleScope.launch {
+            withContext(Dispatchers.IO) {
+                photoViewModel.fetchPhotos().collectLatest {
```
launch this from Viewmodel scope and collect it safely from Fragment scope

atiqaraj replied:
https://medium.com/androiddevelopers/a-safer-way-to-collect-flows-from-android-uis-23080b1f8bda

atiqaraj replied:
https://developer.android.com/topic/libraries/architecture/coroutines

atiqaraj commented on app/src/main/java/com/example/unsplashphotos/ui/adapter/PhotoAdapter.kt:
```diff
@@ -0,0 +1,47 @@
+
+    class PhotoViewHolder(private val binding: ItemPhotoBinding) :
+        RecyclerView.ViewHolder(binding.root) {
+        fun bind(photo: Photo, selectListener: (Photo,FragmentNavigator.Extras) -> Unit) {
```
better to name it `onClickListener`

atiqaraj commented on app/src/main/java/com/example/unsplashphotos/ui/photofullscreen/PhotoFullScreenFragment.kt:
```diff
@@ -0,0 +1,80 @@
+class PhotoFullScreenFragment : Fragment() {
+    private lateinit var binding: FragmentPhotoFullScreenBinding
+    private val photoFullViewModel by viewModels<PhotoFullViewModel>()
+    private var photoId: String = "ph"
```
is this variable required ?

atiqaraj commented on app/src/main/java/com/example/unsplashphotos/ui/photofullscreen/PhotoFullScreenFragment.kt:
```diff
@@ -0,0 +1,80 @@
+
+    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
+        super.onViewCreated(view, savedInstanceState)
+
```
unnecessary line breaks. fix all of those across the project

atiqaraj commented on app/src/main/java/com/example/unsplashphotos/ui/photofullscreen/PhotoFullViewModel.kt:
```diff
@@ -0,0 +1,21 @@
+class PhotoFullViewModel @Inject constructor(private val photoFullScreenUseCase: PhotoFullScreenUseCase) :
+    ViewModel() {
+
+    fun getPhotoById(id: String) = liveData{
```
make it flow instead of livedata

atiqaraj commented on app/src/main/java/com/example/unsplashphotos/ui/photofullscreen/PhotoFullScreenFragment.kt:
```diff
@@ -0,0 +1,80 @@
+                binding.progressBar.visibility = View.GONE
+            } else {
+                binding.progressBar.visibility = View.GONE
+                binding.textTitle.text = "No Data Available"
```
transfer all strings to resource file

atiqaraj commented on app/src/main/java/com/example/unsplashphotos/ui/adapter/PhotoAdapter.kt:
```diff
@@ -0,0 +1,47 @@
+import com.example.unsplashphotos.databinding.ItemPhotoBinding
+
+
+class PhotoAdapter(private val selectListener: (Photo, FragmentNavigator.Extras) -> Unit) : PagingDataAdapter<Photo, PhotoAdapter.PhotoViewHolder>(DIFF_ITEM_CALLBACK) {
```
format the file

atiqaraj commented on app/src/main/java/com/example/unsplashphotos/ui/photofullscreen/PhotoFullScreenFragment.kt:
```diff
@@ -0,0 +1,80 @@
+
+    override fun onResume() {
+        super.onResume()
+        (activity as AppCompatActivity?)!!.supportActionBar!!.hide()
```
use null safe code

atiqaraj commented on app/src/main/java/com/example/unsplashphotos/ui/photofullscreen/PhotoFullScreenFragment.kt:
```diff
@@ -0,0 +1,80 @@
+    override fun onCreateView(
+        inflater: LayoutInflater, container: ViewGroup?,
+        savedInstanceState: Bundle?
+    ): View? {
```
fix return type

atiqaraj commented on app/src/main/java/com/example/unsplashphotos/ui/GalleryFragment.kt:
```diff
@@ -0,0 +1,76 @@
+    private fun setupPhotoRecyclerView() {
+        binding.photosRecyclerView.adapter = photoAdapter
+        photoViewModel
+        val spanCount = 1
```
why span count 1 ?

atiqaraj commented on app/src/main/java/com/example/unsplashphotos/di/NetworkingModule.kt:
```diff
@@ -0,0 +1,45 @@
+package com.example.unsplashphotos.di
+
+import android.app.Application
+import com.example.unsplashphotos.data.RetrofitInstance
```
remove unnecessary imports across the project

atiqaraj commented on app/src/main/java/com/example/unsplashphotos/data/RetrofitInstance.kt:
```diff
@@ -0,0 +1,32 @@
+        }.build()
+
+
+        fun getRetrofitInstance(BASE_URL: String): Retrofit {
```
violation of variable name

atiqaraj commented on app/src/main/java/com/example/unsplashphotos/di/NetworkingModule.kt:
```diff
@@ -0,0 +1,45 @@
+@InstallIn(SingletonComponent::class)
+object NetworkingModule {
+
+    private const val BASE_URL = "https://api.unsplash.com/"
```
transfer this to build file

atiqaraj commented on app/src/main/java/com/example/unsplashphotos/di/NetworkingModule.kt:
```diff
@@ -0,0 +1,45 @@
+        return Retrofit.Builder()
+            .baseUrl(BASE_URL)
+            .client(client)
+            .addConverterFactory(GsonConverterFactory.create(GsonBuilder().create()))
```
provide gson as well

atiqaraj commented on app/src/main/java/com/example/unsplashphotos/data/api/PhotoService.kt:
```diff
@@ -0,0 +1,17 @@
+import retrofit2.http.Query
+
+
+interface PhotoService {
```
in fact rename it to PhotoRemoteDataSource

jhasan95 committed:
refactor: review resolved

- ext.versions
- make ImageLoader singleton using hilt provided context

jhasan95 committed:
refactor: resolve reviews

- add new line
- removed UseCaseModule
- provided gson

jhasan95 committed:
refactor: used viewModelScope in PhotoViewModel and replace livedata with flow in PhotoFullViewModel

jhasan95 committed:
feat: save photo from imageView of fullscreen photo

jhasan95 committed:
feat: download image from download link and save to gallery

jhasan95 committed:
feat: share image from full screen image view

- added FileProvider
- added provider_paths.xml to share access to the External Storage at root folder (path=".") with the name external_files

jhasan95 committed:
refactor: removed unnecessary code

atiqaraj commented on app/src/main/java/com/example/unsplashphotos/data/repository/PhotoPagingSource.kt:
```diff
@@ -0,0 +1,36 @@
+            anchorPage?.prevKey?.plus(1) ?: anchorPage?.nextKey?.minus(1)
+        }
+    }
+}
```
new line

atiqaraj commented on app/src/main/java/com/example/unsplashphotos/ui/ViewUtils.kt:
```diff
@@ -0,0 +1,52 @@
+import java.lang.Exception
+
+
+class ViewUtils() {
```
rename this file

atiqaraj commented on app/src/main/java/com/example/unsplashphotos/ui/ViewUtils.kt:
```diff
@@ -0,0 +1,52 @@
+        fun shareImage(context: Context, photoId: String, imageView: ImageView) {
+
+            val bitmapDrawable: BitmapDrawable = imageView.drawable as BitmapDrawable
+            val bitmap = bitmapDrawable.bitmap
```
extract this function to two method, save to temp and share the image

atiqaraj commented on app/src/main/java/com/example/unsplashphotos/ui/photofullscreen/PhotoFullScreenFragment.kt:
```diff
@@ -0,0 +1,138 @@
+    }
+
+    // Function to check and request permission.
+    private fun checkPermission(permission: String, requestCode: Int) {
```
please follow this to request permission  https://developer.android.com/training/basics/intents/result

atiqaraj commented on app/src/main/java/com/example/unsplashphotos/ui/PhotoViewModel.kt:
```diff
@@ -0,0 +1,29 @@
+
+    fun fetchPhotos() {
+        viewModelScope.launch(Dispatchers.IO) {
+            photos = photoUseCase.fetchPhotos()
```
use cachedIn(viewModelScope) to survive configuration change

atiqaraj replied:
```
@HiltViewModel
class PhotoViewModel @Inject constructor(photoUseCase: PhotoUseCase) : ViewModel() {

    val photos: Flow<PagingData<Photo>> = photoUseCase.fetchPhotos().cachedIn(viewModelScope)

}
```

atiqaraj commented on app/src/main/java/com/example/unsplashphotos/ui/photofullscreen/PhotoFullViewModel.kt:
```diff
@@ -0,0 +1,27 @@
+class PhotoFullViewModel @Inject constructor(private val photoFullScreenUseCase: PhotoFullScreenUseCase) :
+    ViewModel() {
+
+    lateinit var flow: Flow<Photo?>
```
use MutableStateFlow and expose only stateflow. See the difference between a flow and MutableStateFlow

atiqaraj commented on app/src/main/res/layout/activity_main.xml:
```diff
@@ -1,18 +1,20 @@
+            app:layout_constraintBottom_toBottomOf="parent"
+            app:layout_constraintTop_toTopOf="parent" />
+    </androidx.constraintlayout.widget.ConstraintLayout>
+</layout>
```
look this error for other files and fix this

atiqaraj commented on app/src/main/java/com/example/unsplashphotos/ui/ViewUtils.kt:
```diff
@@ -0,0 +1,52 @@
+import android.graphics.Bitmap
+import android.graphics.drawable.BitmapDrawable
+import android.net.Uri
+import android.provider.Settings.Global.getString
```
remove this, find all unused imports and remove them

atiqaraj commented on app/src/main/java/com/example/unsplashphotos/ui/ViewUtils.kt:
```diff
@@ -0,0 +1,52 @@
+
+            val bitmapDrawable: BitmapDrawable = imageView.drawable as BitmapDrawable
+            val bitmap = bitmapDrawable.bitmap
+            var outputStream: FileOutputStream?
```
any reason to define this outside ?

jhasan95 committed:
refactor: renamed ViewUtils to ShareUtils and extracted sharing function into two methods

jhasan95 committed:
refactor: handled permission using registerForActivityResult

jhasan95 committed:
refactor: used cachedIn(viewModelScope) to survive configuration change

jhasan95 committed:
refactor: used MutableStateFlow in PhotoFullViewModel

atiqaraj commented on app/src/main/java/com/example/unsplashphotos/data/repository/DownloaderUtils.kt:
```diff
@@ -0,0 +1,72 @@
+                    DownloadManager.STATUS_RUNNING -> {
+                        Toast.makeText(
+                            context,
+                            "Downloading...",
```
use string resource

atiqaraj commented on app/src/main/java/com/example/unsplashphotos/data/repository/DownloaderUtils.kt:
```diff
@@ -0,0 +1,72 @@
+                        finishDownload = true
+                        Toast.makeText(
+                            context,
+                            "Image downloaded successfully in $directory", Toast.LENGTH_SHORT
```
same

atiqaraj commented on app/src/main/java/com/example/unsplashphotos/data/repository/DownloaderUtils.kt:
```diff
@@ -0,0 +1,72 @@
+                    else -> {
+                        Toast.makeText(
+                            context,
+                            "\"There's nothing to download\"",
```
same

atiqaraj commented on app/src/main/java/com/example/unsplashphotos/data/repository/PhotoPagingSource.kt:
```diff
@@ -0,0 +1,36 @@
+
+
+    override suspend fun load(params: LoadParams<Int>): LoadResult<Int, Photo> {
+
```
unnecessary gap

atiqaraj commented on app/src/main/java/com/example/unsplashphotos/data/repository/PhotoRepoImpl.kt:
```diff
@@ -0,0 +1,57 @@
+    }
+
+    override suspend fun getPhotoById(photoId: String): Photo? {
+
```
same

atiqaraj commented on app/src/main/java/com/example/unsplashphotos/data/repository/PhotoRepoImpl.kt:
```diff
@@ -0,0 +1,57 @@
+        } catch (exception: Exception) {
+            Timber.e(exception.message.toString())
+        }
+
```
please fix this gaps

atiqaraj commented on app/src/main/java/com/example/unsplashphotos/ui/ShareUtils.kt:
```diff
@@ -0,0 +1,52 @@
+
+    companion object {
+        //To share image from the imageView
+        fun shareImage(context: Context, photoId: String, imageView: ImageView) {
```
please share BitmapDrawable instead of imageview

atiqaraj commented on app/src/main/java/com/example/unsplashphotos/ui/photofullscreen/PhotoFullScreenFragment.kt:
```diff
@@ -0,0 +1,112 @@
+    lateinit var downloaderUtils: DownloaderUtils
+    private lateinit var binding: FragmentPhotoFullScreenBinding
+    private val photoFullViewModel by viewModels<PhotoFullViewModel>()
+    private lateinit var photoId: String
```
do we need to store state here ?

jhasan95 committed:
refactor: review resolved

- added string resource

atiqaraj commented on app/src/main/res/navigation/nav_graph.xml:
```diff
@@ -0,0 +1,29 @@
+        app:argType="string" />
+
+    </fragment>
+</navigation>
```
new line

atiqaraj commented on gradle.properties:
```diff
@@ -18,4 +18,8 @@ android.useAndroidX=true
+#Access Key
+CLIENT_ID = "RUvTzcL-6uXdrLTWpudHjLa9MqIvBcjPmJ8bqJwFpQg"
+
+BASE_URL = "https://api.unsplash.com/"
```
it can be removed from here, because if we use this approach we can't use a separate endpoint and api key for the paid version

jhasan95 committed:
refactor: modified PhotoFullViewModel with respect to review

- moved base url and client id to build.gradle

jhasan95 committed:
refactor: removed unnecessary imports

jhasan95 committed:
feat: added GridLayoutManager to show photos in gridview

jhasan95 committed:
refactor: replaced StaggeredGridLayoutManager with GridLayoutManager

jhasan95 committed:
feat: added info fab to show likes count and hide fab, action bar on tap on image

jhasan95 committed:
refactor: reformat files

fahim-choudhury commented on app/src/main/java/com/example/unsplashphotos/ui/ShareUtils.kt:
```diff
@@ -0,0 +1,59 @@
+import java.io.FileOutputStream
+
+
+class ShareUtils() {
```
Can be written using Kotlin's object declaration, you don't need to initialize ShareUtils() by hand. 
Benefits? Singleton by design. Thread-safe. Lazily initialized.

https://kotlinlang.org/docs/object-declarations.html#object-declarations-overview

https://www.raywenderlich.com/23623842-object-in-kotlin-and-the-singleton-pattern

fahim-choudhury commented on app/src/main/java/com/example/unsplashphotos/ui/adapter/PhotoAdapter.kt:
```diff
@@ -0,0 +1,57 @@
+            }
+
+            override fun areContentsTheSame(oldItem: Photo, newItem: Photo): Boolean {
+                return oldItem.id == newItem.id
```
This function shouldn't check the id values, rather check if their contents are the same or not. For now oldItem == newItem should be sufficient.

fahim-choudhury commented on app/src/main/java/com/example/unsplashphotos/domain/usecase/PhotoUseCase.kt:
```diff
@@ -0,0 +1,24 @@
+import javax.inject.Inject
+
+
+class PhotoUseCase @Inject constructor(private val photoPagingSource: PagingSource<Int, Photo>) {
```
Typically, UseCase name should start with a verb that represents some kind of business flow. For example, FetchPhotosUseCase, UpdateUserInfoUseCase etc.
Please read these two articles on UseCases. 
(Good to see that you already have the basic understanding of UseCase. 👍) 

1. https://www.techyourchance.com/how-to-use-case-interactor-java/
2. https://www.techyourchance.com/how-to-use-case-interactor-kotlin/

fahim-choudhury commented on app/src/main/java/com/example/unsplashphotos/domain/usecase/PhotoFullScreenUseCase.kt:
```diff
@@ -0,0 +1,11 @@
+import javax.inject.Inject
+
+
+class PhotoFullScreenUseCase @Inject constructor(private val photoRepo: PhotoRepo) {
```
Please follow the naming convention mentioned in PhotoUseCase comment.

fahim-choudhury commented on app/src/main/java/com/example/unsplashphotos/data/repository/DownloaderUtils.kt:
```diff
@@ -0,0 +1,73 @@
+import com.example.unsplashphotos.R
+import java.io.File
+
+class DownloaderUtils(private val context: Context) {
```
Try to use object declaration mentioned in ShareUtils. Pass context as method parameter instead of constructor parameter.

fahim-choudhury commented on app/src/main/java/com/example/unsplashphotos/domain/usecase/PhotoUseCase.kt:
```diff
@@ -0,0 +1,24 @@
+import androidx.paging.PagingConfig
+import androidx.paging.PagingData
+import androidx.paging.PagingSource
+import com.example.unsplashphotos.data.model.Photo
```
Domain layer UseCases shouldn't have data layer's Photo object. Use different objects for different layers. Use one Photo class with all the annotations required for JSON parsing and database operations and keep it in data layer. (They are called DTO: Data Transfer Object.) 

Use another Photo class containing only the required fields for using in domain and/or presentation layer and keep it in domain layer. Use a mapper for converting to/from these different objects.

See this article for details on how to separate these two types of objects.
https://proandroiddev.com/the-real-repository-pattern-in-android-efba8662b754

fahim-choudhury replied:
Additionally, you can extract an interface from the UseCase that doesn't have any dependency on the data layer. These interfaces will reside on the domain layer. In the data layer, the implementations of these UsesCases will reside, containing dependencies from the data layer (i.e. Retrofit interface or repository). ViewModels will use the domain layer's UseCases and thus the implementation will be abstracted out.

fahim-choudhury commented on app/src/main/java/com/example/unsplashphotos/domain/usecase/PhotoUseCase.kt:
```diff
@@ -0,0 +1,24 @@
+
+    private val config = PagingConfig(pageSize = 20, enablePlaceholders = false)
+
+    fun fetchPhotos(): Flow<PagingData<Photo>> {
```
Android Paging library shouldn't reside inside UseCase. It's place is in data layer. Please refactor accordingly after going through the UseCase articles.

fahim-choudhury reviewed (changes requested):
Please go through the mentioned articles, absorb the content and refactor accordingly.

Great work so far and keep going! 👍👍

jhasan95 committed:
refactor: review resolve

jhasan95 committed:
bug: continuous api call

jhasan95 committed:
refactor: remove line

fahim-choudhury committed:
Fix continuous API calls

jhasan95 committed:
refactor: implemented use case according to the review

jhasan95 committed:
feat: added datastate class

jhasan95 committed:
refactor: rename directory for domain model

jhasan95 committed:
refactor: review resolved

- added network interceptor for
   CLIENT_ID
- refactored dataState to sealed class
- updated gradle to 7.3.3

jhasan95 committed:
refactor: added Links data class for the domain layer and updated mapper accordingly

jhasan95 committed:
refactor: updated mapper for Urls

jhasan95 committed:
refactor: Updated datastate class

jhasan95 committed:
refactor:delete unnecessary class

jhasan95 committed:
feat: added fragment transition animations

jhasan95 committed:
Merge branch 'bug-continuous-api-call' into develop

jhasan95 committed:
docs: Updated README.md

jhasan95 committed:
docs: updated README.md

jhasan95 committed:
docs: updated README.md

jhasan95 committed:
docs: updated README.md

jhasan95 committed:
docs: updated README.md

jhasan95 committed:
docs: updated UML diagram

jhasan95 committed:
docs: Updated README.md

jhasan95 committed:
docs: updated README.md

jhasan95 committed:
docs: updated README.md

jhasan95 committed:
docs: updated README.md

jhasan95 committed:
docs: updated image

Commit:
refactor: Use jetpack compose for the UI(#6)

* feat(Compose): Use jetpack compose in Gallery and Full screen

* refactor: Update composition of gallery screen

* refactor: Update GalleryScreen

* feat(Gallery): Add load and error view

* feat: Convert download feature & permission with compose

* feat: Add info fab on Full Screen

* refactor: Update DI

* refactor: Modify DownloaderUtils

* feat: Add dialog

* feat(Navigation): Add compose destinations library

* feat: Add dialog

* feat: Add screen transition animation

- add fastlane

* feat(lint): Format using ktlint

- Add ktlint

---------

Co-authored-by: Jahid Hasan Naiem <jahid.hasan@monstar-lab.com>

Commit:
feat(Navigation): Introduce Compose destinations library (#7)

Commit:
Update README.md

Commit:
feat: Introduce compose destinations lib (#8)

jhasan95 committed:
refactor: Add image caching with coil compose
